orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +854,306 @@ def _update_onnx_model_initializers(self, state_tensors):
+
+        def _check_optimizer_key_mismatch(training_session_state_dict, state_dict):
+            optimizer = 'optimizer'
+
+            # check for model state key mismatch for the optimizer state_dict
+            unexpected_optimizer_model_state_keys = list(set(state_dict[optimizer]) - \
+                set(training_session_state_dict[optimizer]))
+            if len(unexpected_optimizer_model_state_keys) > 0:
+                raise RuntimeError("Unexpected keys: {} in state_dict[optimizer]"\
+                    .format(unexpected_optimizer_model_state_keys))
+            missing_optimizer_model_state_keys = list(set(training_session_state_dict[optimizer]) - \
+                set(state_dict[optimizer]))
+            if len(missing_optimizer_model_state_keys) > 0:
+                raise RuntimeError("Missing keys: {} in state_dict[optimizer]"\
+                    .format(missing_optimizer_model_state_keys))
+
+            # check for optimizer state keys mismatch
+            for model_state_key in training_session_state_dict[optimizer]:
+                training_session_optimizer_state_keys = set(training_session_state_dict[optimizer][model_state_key])
+                state_dict_optimizer_state_keys = set(state_dict[optimizer][model_state_key])
+                unexpected_keys = list(state_dict_optimizer_state_keys - training_session_optimizer_state_keys)
+                if len(unexpected_keys):
+                    raise RuntimeError("Unexpected keys: {} in state_dict[optimizer][{}]"\
+                        .format(unexpected_keys, model_state_key))
+                missing_keys = list(training_session_optimizer_state_keys - state_dict_optimizer_state_keys)
+                if len(missing_keys):
+                    raise RuntimeError("Missing keys: {} in state_dict[optimizer][{}]"\
+                        .format(missing_keys, model_state_key))
+
+        def _check_key_mismatch(training_session_state_dict, state_dict):
+            model = 'model'
+            optimizer = 'optimizer'
+            # check presence of 'model' in the input state_dict
+            if model not in state_dict:
+                raise RuntimeError("Missing key: model in state_dict")
+            # check presence of 'optimizer' in the input state_dict
+            if optimizer not in state_dict:
+                raise RuntimeError("Missing key: optimizer in state_dict")
+
+            _check_model_key_mismatch(training_session_state_dict, state_dict)
+
+            _check_optimizer_key_mismatch(training_session_state_dict, state_dict)
+
+
+        # extract state dict from the current training session. this is to persist the states between
+        # two training sessions.
+        # for example, if user provided only the model states, the optimizer states from the current
+        # training session must be persisted
+        states = {}
```

Comment:
can we name the 'states' as 'current_state' or something more clear? it could be confusing to understand the purpose/distinction of 'state' v/s 'state_dict'